File: packages/core/src/search/query.ts
import { loadConfigAsync } from '../config/config.js';
import { getDb } from '../db/index.js';
import { embedQuery } from '../gemini/embeddings.js';

/**
 * Search filters for vector similarity search.
 */
export interface SearchFilters {
  /** Filter by repos.primary_language (case-insensitive) */
  language?: string;
  /** Filter by repos.is_starred = true */
  starred?: boolean;
  /** Filter by repos.is_owned = true */
  owned?: boolean;
  /** Filter by repos.owner (case-insensitive) */
  owner?: string;
  /** Number of results to return, default 3 */
  limit?: number;
}

/**
 * A single search result with repository and chunk information.
 */
export interface SearchResult {
  repoFullName: string;
  owner: string;
  repoName: string;
  description: string | null;
  language: string | null;
  stars: number;
  isStarred: boolean;
  isOwned: boolean;
  htmlUrl: string;
  chunkType: 'metadata' | 'readme' | 'file_tree';
  content: string;
  /** Cosine similarity score, 0-1 */
  similarity: number;
}

/**
 * Result of a search query with timing metrics.
 */
export interface SearchQueryResult {
  results: SearchResult[];
  /** How many chunks were searched before deduplication */
  totalConsidered: number;
  /** Time taken to embed the query in milliseconds */
  queryEmbeddingMs: number;
  /** Time taken to run the vector search in milliseconds */
  searchMs: number;
}

/**
 * Raw row returned from the vector similarity query.
 */
interface RawSearchRow {
  full_name: string;
  owner: string;
  name: string;
  description: string | null;
  primary_language: string | null;
  stars: number;
  is_starred: number;
  is_owned: number;
  html_url: string;
  chunk_type: string;
  content: string;
  distance: number;
}

/**
 * Empty result to return on errors.
 */
function emptyResult(): SearchQueryResult {
  return {
    results: [],
    totalConsidered: 0,
    queryEmbeddingMs: 0,
    searchMs: 0,
  };
}

/**
 * Search repositories using vector similarity.
 *
 * This function:
 * 1. Embeds the query string using Gemini
 * 2. Runs a vector similarity search using sqlite-vec
 * 3. Applies optional filters (language, starred, owned, owner)
 * 4. Deduplicates results by repository, keeping highest similarity chunk
 * 5. Returns results with timing metrics
 *
 * @param query - The natural language search query
 * @param filters - Optional filters to apply
 * @returns Search results with timing metrics, never throws
 */
export async function searchRepos(
  query: string,
  filters: SearchFilters = {}
): Promise<SearchQueryResult> {
  try {
    // Load config
    const config = await loadConfigAsync();
    if (!config.geminiKey) {
      return emptyResult();
    }

    // Embed the query
    const embedStart = performance.now();
    const embedding = await embedQuery(config.geminiKey, query);
    const queryEmbeddingMs = performance.now() - embedStart;

    if (!embedding) {
      return {
        results: [],
        totalConsidered: 0,
        queryEmbeddingMs,
        searchMs: 0,
      };
    }

    // Get database connection (sqlite-vec is loaded automatically via migrate)
    const db = getDb(config.dbPath);

    // Serialize embedding as Buffer for sqlite-vec
    const embeddingBuffer = Buffer.from(new Float32Array(embedding).buffer);

    // Build dynamic WHERE clause and params
    const whereClauses: string[] = [];
    const params: (string | number | Buffer)[] = [embeddingBuffer];

    if (filters.language !== undefined) {
      whereClauses.push('LOWER(r.primary_language) = LOWER(?)');
      params.push(filters.language);
    }

    if (filters.starred === true) {
      whereClauses.push('r.is_starred = 1');
    }

    if (filters.owned === true) {
      whereClauses.push('r.is_owned = 1');
    }

    if (filters.owner !== undefined) {
      whereClauses.push('LOWER(r.owner) = LOWER(?)');
      params.push(filters.owner);
    }

    // Default limit is 3, but we fetch more to allow for deduplication
    const limit = filters.limit ?? 3;
    // Fetch extra rows to ensure we have enough after deduplication
    const fetchLimit = limit * 5;
    params.push(fetchLimit);

    // Build the SQL query
    const whereClause =
      whereClauses.length > 0 ? `AND ${whereClauses.join(' AND ')}` : '';

    const sql = `
      SELECT
        r.full_name,
        r.owner,
        r.name,
        r.description,
        r.primary_language,
        r.stars,
        r.is_starred,
        r.is_owned,
        r.url as html_url,
        c.chunk_type,
        c.content,
        vec_distance_cosine(ce.embedding, ?) as distance
      FROM chunk_embeddings ce
      JOIN chunks c ON c.id = ce.chunk_id
      JOIN repos r ON r.id = c.repo_id
      WHERE 1=1
        ${whereClause}
      ORDER BY distance ASC
      LIMIT ?
    `;

    // Execute the search
    const searchStart = performance.now();
    const rows = db.prepare(sql).all(...params) as RawSearchRow[];
    const searchMs = performance.now() - searchStart;

    const totalConsidered = rows.length;

    // Deduplicate by full_name, keeping highest similarity (lowest distance)
    const seenRepos = new Map<string, SearchResult>();
    for (const row of rows) {
      if (!seenRepos.has(row.full_name)) {
        seenRepos.set(row.full_name, {
          repoFullName: row.full_name,
          owner: row.owner,
          repoName: row.name,
          description: row.description,
          language: row.primary_language,
          stars: row.stars,
          isStarred: row.is_starred === 1,
          isOwned: row.is_owned === 1,
          htmlUrl: row.html_url,
          chunkType: row.chunk_type as 'metadata' | 'readme' | 'file_tree',
          content: row.content,
          similarity: 1 - row.distance,
        });
      }
    }

    // Convert to array and limit to requested count
    const results = Array.from(seenRepos.values()).slice(0, limit);

    return {
      results,
      totalConsidered,
      queryEmbeddingMs,
      searchMs,
    };
  } catch {
    // Never throw - return empty results on any error
    return emptyResult();
  }
}

File: internal/search/query.go
// Package search handles vector similarity search for repositories.
package search

import (
	"context"
	"database/sql"
	"sort"
	"strings"
	"time"

	"github.com/hackastak/repog/internal/provider"
)

// SearchFilters contains optional filters for search queries.
type SearchFilters struct {
	Language *string
	Starred  *bool
	Owned    *bool
	Owner    *string
	Limit    int // default 3
}

// SearchResult represents a single search result.
type SearchResult struct {
	RepoFullName string
	Owner        string
	RepoName     string
	Description  string
	Language     string
	Stars        int
	IsStarred    bool
	IsOwned      bool
	HTMLURL      string
	ChunkType    string
	Content      string
	Similarity   float64
}

// SearchQueryResult contains the search results and timing metrics.
type SearchQueryResult struct {
	Results          []SearchResult
	TotalConsidered  int
	QueryEmbeddingMs int64
	SearchMs         int64
}

// SearchRepos embeds the query and performs a cosine similarity search
// against chunk_embeddings. Results are deduplicated by repo (highest
// similarity chunk per repo), then capped at Limit.
func SearchRepos(ctx context.Context, db *sql.DB, embedProvider provider.EmbeddingProvider, query string, filters SearchFilters) (SearchQueryResult, error) {
	result := SearchQueryResult{
		Results: make([]SearchResult, 0),
	}

	// Set default limit
	limit := filters.Limit
	if limit <= 0 {
		limit = 3
	}

	// Embed the query
	embedStart := time.Now()
	embedding := embedProvider.EmbedQuery(ctx, query)
	result.QueryEmbeddingMs = time.Since(embedStart).Milliseconds()

	if embedding == nil {
		return result, nil
	}

	// Convert embedding to bytes for sqlite-vec
	embeddingBytes := provider.Float32SliceToBytes(embedding)

	// Build dynamic WHERE clause
	var whereClauses []string
	var args []interface{}
	args = append(args, embeddingBytes)

	if filters.Language != nil {
		whereClauses = append(whereClauses, "LOWER(r.language) = LOWER(?)")
		args = append(args, *filters.Language)
	}

	if filters.Starred != nil && *filters.Starred {
		whereClauses = append(whereClauses, "r.is_starred = 1")
	}

	if filters.Owned != nil && *filters.Owned {
		whereClauses = append(whereClauses, "r.is_owned = 1")
	}

	if filters.Owner != nil {
		whereClauses = append(whereClauses, "LOWER(r.owner) = LOWER(?)")
		args = append(args, *filters.Owner)
	}

	whereClause := ""
	if len(whereClauses) > 0 {
		whereClause = "AND " + strings.Join(whereClauses, " AND ")
	}

	// Fetch extra rows for deduplication
	fetchLimit := limit * 5
	args = append(args, fetchLimit)

	sqlQuery := `
		SELECT
			r.full_name,
			r.owner,
			r.name,
			COALESCE(r.description, ''),
			COALESCE(r.language, ''),
			r.stars,
			r.is_starred,
			r.is_owned,
			r.html_url,
			c.chunk_type,
			c.content,
			vec_distance_cosine(ce.embedding, ?) as distance
		FROM chunk_embeddings ce
		JOIN chunks c ON c.id = ce.chunk_id
		JOIN repos r ON r.id = c.repo_id
		WHERE 1=1 ` + whereClause + `
		ORDER BY distance ASC
		LIMIT ?
	`

	// Execute search
	searchStart := time.Now()
	rows, err := db.QueryContext(ctx, sqlQuery, args...)
	result.SearchMs = time.Since(searchStart).Milliseconds()

	if err != nil {
		return result, err
	}
	defer func() { _ = rows.Close() }()

	// Collect results and deduplicate by repo+chunk_type (keep best of each type per repo)
	// Key: "repo_full_name|chunk_type"
	seenChunks := make(map[string]SearchResult)
	// Track best similarity per repo for sorting
	bestRepoSimilarity := make(map[string]float64)

	for rows.Next() {
		var r SearchResult
		var isStarred, isOwned int
		var distance float64

		err := rows.Scan(
			&r.RepoFullName, &r.Owner, &r.RepoName, &r.Description,
			&r.Language, &r.Stars, &isStarred, &isOwned,
			&r.HTMLURL, &r.ChunkType, &r.Content, &distance,
		)
		if err != nil {
			continue
		}

		r.IsStarred = isStarred == 1
		r.IsOwned = isOwned == 1
		r.Similarity = 1 - distance

		result.TotalConsidered++

		// Keep highest similarity chunk of each type per repo
		key := r.RepoFullName + "|" + r.ChunkType
		if existing, exists := seenChunks[key]; !exists || r.Similarity > existing.Similarity {
			seenChunks[key] = r
		}

		// Track best similarity for this repo
		if r.Similarity > bestRepoSimilarity[r.RepoFullName] {
			bestRepoSimilarity[r.RepoFullName] = r.Similarity
		}
	}

	// Group chunks by repo, sort repos by best similarity, then cap at limit repos
	repoChunks := make(map[string][]SearchResult)
	for _, r := range seenChunks {
		repoChunks[r.RepoFullName] = append(repoChunks[r.RepoFullName], r)
	}

	// Sort repos by best similarity
	repos := make([]string, 0, len(repoChunks))
	for repo := range repoChunks {
		repos = append(repos, repo)
	}
	sort.Slice(repos, func(i, j int) bool {
		return bestRepoSimilarity[repos[i]] > bestRepoSimilarity[repos[j]]
	})

	// Cap at limit repos, then flatten all chunks from those repos
	if len(repos) > limit {
		repos = repos[:limit]
	}
	for _, repo := range repos {
		// Sort chunks within repo: metadata first, then readme, then file_tree
		chunks := repoChunks[repo]
		sort.Slice(chunks, func(i, j int) bool {
			order := map[string]int{"metadata": 0, "readme": 1, "file_tree": 2}
			return order[chunks[i].ChunkType] < order[chunks[j].ChunkType]
		})
		result.Results = append(result.Results, chunks...)
	}

	return result, nil
}
